src/h3lib/lib/h3Index.c

```diff
@@ -779,6 +779,88 @@ void H3_EXPORT(h3ToGeoBoundary)(H3Index h3, GeoBoundary* gb) {
                           H3_EXPORT(h3IsPentagon)(h3), gb);
 }
 
+/**
+ * Returns the max number of possible icosahedron faces an H3 index
+ * may intersect.
+ *
+ * @return int count of faces
+ */
+int H3_EXPORT(maxFaceCount)(H3Index h3) {
+    // a pentagon always intersects 5 faces, a hexagon never intersects more
+    // than 2 (but may only intersect 1)
+    return H3_EXPORT(h3IsPentagon)(h3) ? 5 : 2;
+}
+
+/**
+ * Find all icosahedron faces intersected by a given  H3 index, represented
+ * as integers from 0-19. The array is sparse; since 0 is a valid value,
+ * invalid array values are represented as -1. It is the responsibility of
+ * the caller to filter out invalid values.
+ *
+ * @param h3 The H3 index
+ * @param out Output array. Must be of size maxFaceCount(h3).
+ */
+void H3_EXPORT(h3GetFaces)(H3Index h3, int* out) {
+    int res = H3_GET_RESOLUTION(h3);
+    int isPentagon = H3_EXPORT(h3IsPentagon)(h3);
+
+    // We can't use the vertex-based approach here for class II pentagons,
+    // because all their vertices are on the icosahedron edges. Their
+    // direct child pentagons cross the same faces, so use those instead.
+    if (isPentagon && !isResClassIII(res)) {
+        H3Index childPentagon = makeDirectChild(h3, 0);
+        H3_EXPORT(h3GetFaces)(childPentagon, out);
+        return;
+    }
+
+    // convert to FaceIJK
+    FaceIJK fijk;
+    _h3ToFaceIjk(h3, &fijk);
+
+    // Get all vertices as FaceIJK addresses. For simplicity, always
+    // initialize the array with 6 verts, ignoring the last one for pentagons
+    FaceIJK* fijkVerts = malloc(NUM_HEX_VERTS * sizeof(FaceIJK));
+    int vertexCount;
+
+    if (isPentagon) {
+        vertexCount = NUM_PENT_VERTS;
+        _faceIjkPentToVerts(&fijk, &res, fijkVerts);
+    } else {
+        vertexCount = NUM_HEX_VERTS;
+        _faceIjkToVerts(&fijk, &res, fijkVerts);
+    }
+
+    // We may not use all of the slots in the output array,
+    // so fill with invalid values to indicate unused slots
+    int faceCount = H3_EXPORT(maxFaceCount)(h3);
+    for (int i = 0; i < faceCount; i++) {
+        out[i] = INVALID_FACE;
+    }
+
+    // add each vertex face, using the output array as a hash set
+    for (int i = 0; i < vertexCount; i++) {
+        FaceIJK* vert = &fijkVerts[i];
+
+        // Adjust overage, determining whether this vertex is
+        // on another face
+        if (isPentagon) {
+            _adjustPentVertOverage(vert, res);
+        } else {
+            _adjustOverageClassII(vert, res, 0, 1);
+        }
+
+        // Save the face to the output array
+        int face = vert->face;
+        int pos = face % faceCount;
```

Comment:
Is this hashing approach being chosen for a specific reason?

Reply:
It was easy? And I had a model for how to do it in the `kRing` implementation. Is there another option you'd prefer?

Reply:
I personally don't see a problem with it, but it's probably the same amount of compute effort (or slightly more for edge-intersecting hexagons) to just scan the array for the first `INVALID_FACE` to write the value, or abort the scan if the value is found in the array.

Reply:
Simplified the logic here to iterate over the array, removing modulo logic. In practice, the array is never longer than 5 elements, so it shouldn't make a difference.